src/validation/property-validator-schema.js

```diff
@@ -0,0 +1,260 @@
+  .when(
+    joi.object().unknown().keys({ type: 'datetime' }),
+    { then: makeTypeConstraintsSchema('datetime') })
+  .when(
+    joi.object().unknown().keys({ type: 'date' }),
+    { then: makeTypeConstraintsSchema('date') })
+  .when(
+    joi.object().unknown().keys({ type: 'enum' }),
+    { then: makeTypeConstraintsSchema('enum') })
+  .when(
+    joi.object().unknown().keys({ type: 'uuid' }),
+    { then: makeTypeConstraintsSchema('uuid') })
+  .when(
+    joi.object().unknown().keys({ type: 'attachmentReference' }),
+    { then: makeTypeConstraintsSchema('attachmentReference') })
+  .when(
+    joi.object().unknown().keys({ type: 'array' }),
+    { then: makeTypeConstraintsSchema('array') })
+  .when(
+    joi.object().unknown().keys({ type: 'object' }),
+    { then: makeTypeConstraintsSchema('object') })
+  .when(
+    joi.object().unknown().keys({ type: 'hashtable' }),
+    { then: makeTypeConstraintsSchema('hashtable') })
+  .when(
+    joi.object().unknown().keys({ type: joi.func() }),
+    { then: joi.object().unknown() });
+
+/**
+ * A partial schema for a single entry in a "propertyValidators" object at either the top level of a document definition
+ * or nested within an "object" validator.
+ */
+module.exports = exports = schema;
+
+// Defined as a function rather than a plain object because it contains lazy references that result in recursive
+// references between the complex types (e.g. "array", "object", "hashtable") and the main "propertyValidators" schema
+function typeSpecificConstraintSchemas() {
+  return {
+    string: {
+      mustNotBeEmpty: dynamicConstraintSchema(joi.boolean()),
+      regexPattern: dynamicConstraintSchema(regexSchema),
+      minimumLength: dynamicConstraintSchema(integerSchema.min(0)),
+      maximumLength: maximumSizeConstraintSchema('minimumLength'),
+      minimumValue: dynamicConstraintSchema(joi.string()),
+      minimumValueExclusive: dynamicConstraintSchema(joi.string()),
+      maximumValue: dynamicConstraintSchema(joi.string()),
+      maximumValueExclusive: dynamicConstraintSchema(joi.string())
+    },
+    integer: {
+      minimumValue: dynamicConstraintSchema(integerSchema),
+      minimumValueExclusive: dynamicConstraintSchema(integerSchema),
+      maximumValue: maximumValueInclusiveNumberConstraintSchema(integerSchema),
+      maximumValueExclusive: maximumValueExclusiveNumberConstraintSchema(integerSchema)
+    },
+    float: {
+      minimumValue: dynamicConstraintSchema(joi.number()),
+      minimumValueExclusive: dynamicConstraintSchema(joi.number()),
+      maximumValue: maximumValueInclusiveNumberConstraintSchema(joi.number()),
+      maximumValueExclusive: maximumValueExclusiveNumberConstraintSchema(joi.number())
+    },
+    boolean: { },
+    datetime: {
+      minimumValue: dynamicConstraintSchema(datetimeSchema),
+      minimumValueExclusive: dynamicConstraintSchema(datetimeSchema),
+      maximumValue: dynamicConstraintSchema(datetimeSchema),
```

Comment:
Joi only exposes range constraints (e.g. `greater`, `less`, etc.) for numeric types, so other types have to do without.

Reply:
Correction: Joi's [date](https://github.com/hapijs/joi/blob/v13.1.1/API.md#date---inherits-from-any) type _does_ support `min` and `max` values (but not `greater` or `less`). I will investigate if it makes sense to use them.

Reply:
Er... now I remember why I chose to use a string with a regex rather than [`joi.date()`](https://github.com/hapijs/joi/blob/v13.1.1/API.md#date---inherits-from-any) for the `date` and `datetime` data types: when Joi is running, it is doing so in a Node.js process, which supports a different subset of the ISO 8601 date-time string format than Sync Gateway's JavaScript interpreter does. It ends up allowing date strings that cannot be parsed when the generated sync function is used in Sync Gateway. I'll put a comment in the code to explain this.

Reply:
👍